tmpl/blog/mirage-entropy.md

```diff
@@ -0,0 +1,341 @@
+state, just by looking at their outputs. In a sense, someone trying to predict
+the outputs can not take a shortcut and is reduced to starting the generator
+with all the possible states and checking if the output of their generator
+matches the observed sequence, a laborious task. This is how we can quantify a
+CSPRNG unpredictability: it takes trying about half of all the possible states to
+guess the state.
+
+MirageOS's security stack contains a CSPRNG, a design called [Fortuna][fortuna].
+What it really does, is encrypt the simple sequence `0, 1, 2, 3, ...` with AES
+(AES-CTR) using a secret key. This makes it as resistant to prediction as AES is
+to [known-plaintext attacks][wiki-known-plaintext]. After each output, it
+generates a bit more, hashes that, and uses the result as the next key. This is
+not to improve the statistical randomness, as this is already guaranteed by AES.
+Rather, it's a form of [forward secrecy][wiki-forward-secrecy]: an attacker who
+learns the secret key at some point would need to perform the [preimage
+attack][wiki-preimage-attack] on the hash function to figure out the earlier key
+and reconstruct the earlier outputs.
+
+[wiki-random-sequence]: https://en.wikipedia.org/wiki/Random_sequence
+[wiki-statistical-randomness]: https://en.wikipedia.org/wiki/Statistical_randomness
+[diehard-web]: http://www.stat.fsu.edu/pub/diehard/
+[marsaglia-pi-randomness]: http://interstat.statjournals.net/YEAR/2005/articles/0510005.pdf
+[wiki-algorithmic-randomness]: https://en.wikipedia.org/wiki/Algorithmically_random_sequence
+[wolfram-rule-30]: http://mathworld.wolfram.com/Rule30.html
+[mersenne-twister]: https://en.wikipedia.org/wiki/Mersenne_Twister
+[dgrb-boskovic]: http://qrbg.irb.hr/
+[fortuna]: https://www.schneier.com/fortuna.html
+[wiki-known-plaintext]: https://en.wikipedia.org/wiki/Known-plaintext_attack
+[wiki-forward-secrecy]: https://en.wikipedia.org/wiki/Forward_secrecy
+[wiki-preimage-attack]: https://en.wikipedia.org/wiki/Preimage_attack
+
+## Entropy
+
+Still, just like any other software RNG, Fortuna is just a deterministic
+algorithm. Its entire output is as unpredictable as its initial seed. From the
+information perspective, a RNG can only stretch what was unpredictable about its
+initial seed into an equally unpredictable sequence. The best generators do not
+give out more hints about their staring position, but they can never out-race
+the amount of unpredictability that was contained in the seed.
+
+We often call this unpredictable bit *entropy*, after the quality we are hoping
+it can bring into our RNG - the actual unpredictability. In a sense, by
+employing an algorithmic generator, we have just shifted the burden of being
+unpredictable around. But now we're cornered and have to search for it in the
+only place where a computer can find it -- in the physical world.
+
+A typical (kernel-level) RNG reaches out into the world around it through
```

Comment:
Do they call this `RNG` ? Otherwise I would call it differently (e.g. entropy provider) to minimize confusion. In any case as mentioned above you should use `PRNG` were appropriate so that we can make a clear distinction between the pseudo random numbers and the entropy random numbers.